workerextension.go

```diff
@@ -27,154 +24,73 @@ import (
-				continue
-			}
-
-			var ok bool
-			if fc, ok = fromContext(fr.Context()); !ok {
-				continue
-			}
-		}
-
-		fc.worker = w
-
-		fc.responseWriter = rq.Response
-		fc.handlerParameters = rq.CallbackParameters
-
-		// Queue the request and wait for completion if Done channel was provided
-		logger.LogAttrs(context.Background(), slog.LevelInfo, "queue the external worker request", slog.String("worker", w.name), slog.Int("thread", thread.threadIndex))
-
-		w.requestChan <- fc
-		if rq.AfterFunc != nil {
-			go func() {
-				<-fc.done
-
-				if rq.AfterFunc != nil {
-					rq.AfterFunc(fc.handlerReturn)
-				}
-			}()
-		}
-	}
-}
+// EXPERIMENTAL: SendRequest sends an HTTP request to the worker and writes the response to the provided ResponseWriter.
+func (w Worker) SendRequest(rw http.ResponseWriter, r *http.Request) error {
+	worker := getWorkerByName(w.name)
 
-// EXPERIMENTAL: NewWorker creates a Worker instance to embed in a custom struct implementing the Worker interface.
-// The returned instance may be sufficient on its own for simple use cases.
-func NewWorker(name, fileName string, minThreads int, env PreparedEnv) Worker {
-	return &defaultWorker{
-		name:           name,
-		fileName:       fileName,
-		env:            env,
-		minThreads:     minThreads,
-		requestChan:    make(chan *WorkerRequest),
-		activatedCount: atomic.Int32{},
-		drainCount:     atomic.Int32{},
+	if worker == nil {
+		return errors.New("worker not found: " + w.name)
 	}
-}
 
-type defaultWorker struct {
-	name           string
-	fileName       string
-	env            PreparedEnv
-	minThreads     int
-	requestChan    chan *WorkerRequest
-	activatedCount atomic.Int32
-	drainCount     atomic.Int32
-}
+	fr, err := NewRequestWithContext(
+		r,
+		WithOriginalRequest(r),
+		WithWorkerName(w.name),
+	)
 
-func (w *defaultWorker) Name() string {
-	return w.name
-}
+	if err != nil {
+		return err
+	}
 
-func (w *defaultWorker) FileName() string {
-	return w.fileName
-}
+	err = ServeHTTP(rw, fr)
 
-func (w *defaultWorker) Env() PreparedEnv {
-	return w.env
-}
+	if err != nil {
+		return err
+	}
 
-func (w *defaultWorker) MinThreads() int {
-	return w.minThreads
+	return nil
 }
 
-func (w *defaultWorker) OnReady(_ int) {
-	w.activatedCount.Add(1)
-}
+// EXPERIMENTAL: SendMessage sends a message to the worker and waits for a response.
+func (w Worker) SendMessage(message any, rw http.ResponseWriter) (any, error) {
+	internalWorker := getWorkerByName(w.name)
 
-func (w *defaultWorker) OnShutdown(_ int) {
-	w.drainCount.Add(1)
-}
+	if internalWorker == nil {
+		return nil, errors.New("worker not found: " + w.name)
+	}
 
-func (w *defaultWorker) OnServerShutdown(_ int) {
-	w.drainCount.Add(-1)
-	w.activatedCount.Add(-1)
-}
+	fc := newFrankenPHPContext()
+	fc.logger = logger
+	fc.worker = internalWorker
+	fc.responseWriter = rw
+	fc.handlerParameters = message
+
+	internalWorker.handleRequest(fc)
```

Comment:
Previously, there was a way to signal to the external worker that we were ready to receive requests, but now it only applies if the user knew to pass an option in. Can we safely handle the case where we start receiving messages/requests before the worker is ready?